cmd/common/registry.go

```diff
@@ -0,0 +1,377 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package common
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"os"
+	"sort"
+	"strings"
+	"time"
+
+	log "github.com/sirupsen/logrus"
+	"github.com/spf13/cobra"
+
+	"github.com/inspektor-gadget/inspektor-gadget/cmd/common/frontends/console"
+	"github.com/inspektor-gadget/inspektor-gadget/cmd/common/utils"
+	cols "github.com/inspektor-gadget/inspektor-gadget/pkg/columns"
+	gadgetcontext "github.com/inspektor-gadget/inspektor-gadget/pkg/gadget-context"
+	gadgetregistry "github.com/inspektor-gadget/inspektor-gadget/pkg/gadget-registry"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/gadgets"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/logger"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/operators"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/params"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/parser"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/runtime"
+)
+
+// AddCommandsFromRegistry adds all gadgets known by the registry as cobra commands as a subcommand to their categories
+func AddCommandsFromRegistry(rootCmd *cobra.Command, runtime runtime.Runtime, columnFilters []cols.ColumnFilter) {
+	runtimeParams := runtime.GlobalParamDescs().ToParams()
+
+	// Build lookup
+	lookup := make(map[string]*cobra.Command)
+	for _, cmd := range rootCmd.Commands() {
+		lookup[cmd.Name()] = cmd
+	}
+
+	// Add runtime flags
+	addFlags(rootCmd, runtimeParams)
+
+	// Add operator global flags
+	operatorsParamsCollection := operators.GlobalParamsCollection()
+	for _, operatorParams := range operatorsParamsCollection {
+		addFlags(rootCmd, operatorParams)
+	}
+
+	// Add all known gadgets to cobra in their respective categories
+	categories := gadgets.GetCategories()
+	for _, gadgetDesc := range gadgetregistry.GetAll() {
+		cmd, ok := lookup[gadgetDesc.Category()]
+		if !ok {
+			// Category not found, add it
+			categoryDescription, ok := categories[gadgetDesc.Category()]
+			if !ok {
+				panic(fmt.Errorf("category unknown: %q", gadgetDesc.Category()))
+			}
+			cmd = &cobra.Command{
+				Use:   gadgetDesc.Category(),
+				Short: categoryDescription,
+			}
+			rootCmd.AddCommand(cmd)
+			lookup[gadgetDesc.Category()] = cmd
+		}
+		cmd.AddCommand(buildCommandFromGadget(gadgetDesc, columnFilters, runtime, runtimeParams, operatorsParamsCollection))
+	}
+}
+
+func buildGadgetDoc(gadgetDesc gadgets.GadgetDesc, parser parser.Parser) string {
+	var out strings.Builder
+	fmt.Fprintf(&out, "%s\n\n", gadgetDesc.Description())
+
+	if parser != nil {
+		fmt.Fprint(&out, "Available columns:\n")
+		for columnName, description := range parser.GetColumnNamesAndDescription() {
+			fmt.Fprintf(&out, "\t%s\n", columnName)
+			if description != "" {
+				fmt.Fprintf(&out, "\t  %s\n", description)
+			}
+		}
+	}
+	return out.String()
+}
```

Comment:
I'm wondering if this is really the best place to print the available columns. I feel it's better to print them in the description of the `--columns` flag?

Current:

```bash
$ ./local-gadget trace exec -h
Trace new processes

Available columns:
	comm
	ret
	args
	ppid
	uid
	mntns
	container
	timestamp
	pid

Usage:
  local-gadget trace exec [flags]

Flags:
  -C, --columns strings        Columns to output (default [container,pid,ppid,comm,ret,args])
  -c, --containername string   Show only data from containers with that name (default "")
  -F, --filter strings         Filter rules
  -h, --help                   help for exec
  -o, --output string          Output format (json, columns).
                               
                               Columns (columns)
                                 The output of the gadget is formatted in human readable columns
                               JSON (json)
                                 The output of the gadget is returned as raw JSON
                               
                                (default "columns")
  -t, --timeout int            Number of seconds that the gadget will run for, 0 to disable
  -v, --verbose                Print debug information

Global Flags:
  -r, --runtimes string   Container runtimes to be used separated by comma. Supported values are: docker, containerd, cri-o (default "docker,containerd,cri-o")

```


Proposed in https://github.com/inspektor-gadget/inspektor-gadget/commit/4e3259a36e475851bfac52cbfda89030d2e31c93

```
$ ./local-gadget trace exec -h
Trace new processes

Usage:
  local-gadget trace exec [flags]

Flags:
  -C, --columns strings        Columns to output
                               
                               Available columns:
                               	timestamp
                               	ppid
                               	comm
                               	ret
                               	args
                               	mntns
                               	container
                               	pid
                               	uid
                                (default [container,pid,ppid,comm,ret,args])
  -c, --containername string   Show only data from containers with that name (default "")
  -F, --filter strings         Filter rules
  -h, --help                   help for exec
  -o, --output string          Output format (json, columns).
                               
                               Columns (columns)
                                 The output of the gadget is formatted in human readable columns
                               JSON (json)
                                 The output of the gadget is returned as raw JSON
                               
                                (default "columns")
  -t, --timeout int            Number of seconds that the gadget will run for, 0 to disable
  -v, --verbose                Print debug information

Global Flags:
  -r, --runtimes string   Container runtimes to be used separated by comma. Supported values are: docker, containerd, cri-o (default "docker,containerd,cri-o")
```

Reply:
I'm not sure if we already discussed this. Do we want to have two different flags to define the output and which columns, instead of `-o columns=a,b,c`? I don't like that flag `-C` has no sense when `-o json` is used. What should we do in that case? Ignore `-C`, return an error? Having them in the same flag avoids that dilemma.

Reply:
> I'm not sure if we already discussed this. Do we want to have two different flags to define the output and which columns, instead of `-o columns=a,b,c`? I don't like that flag `-C` has no sense when `-o json` is used. What should we do in that case? Ignore `-C`, return an error? Having them in the same flag avoids that dilemma.

I don't think it's a big issue to have `-C` not being used in every case, depending on other arguments - I'm sure most command line tools that are a bit more complex have the same issue. I don't have a strong opinion on which way to use, but I think a dedicated argument is more clear in its usage and also more helpful when using autocompletion.

If we go forward with the dedicated `-C`, I wouldn't error out in case you set `-C` but use JSON output for example, mainly because I often run into situations where I build a long command and then want to change a single thing, like the output. Erroring out would force me into removing specific parameters to just get it to run again. I think it would be more punishing than helping in most situations :)

Reply:
Ah, I haven't noticed this change. IMO it's better to only have `-o` and control all the output modes through it.  I think it's more clear and causes less confusion as there is a simple flag to control it and we don't have to think about the `-C` combined with `-o != columns` cases. It also seems to be more aligned with `kubectl`.

Reply:
Okay, I changed it.

Reply:
```
GetColumnNamesAndDescription() map[string]string
```
Since this uses a map instead of an array, the order of the fields in the "Available columns" section is not specified. It should either use a deterministic order (either alphabetical or logical).

Reply:
This is sorted alphabetically now.